### PR TITLE
Add async ssa graph executor communicator

### PR DESCRIPTION
**Pull request by jacquesqiao**

### CODE REVIEW

File: paddle/fluid/framework/details/multi_devices_graph_pass.h [hunk 1]
```diff
@@ -115,6 +115,32 @@ class AllReduceSSAGraphBuilder : public MultiDevSSAGraphBuilderBase {
...
+                          const std::string &g_name) const override {}
+
+  bool NeedCollectiveForGrad(const std::string &grad_name,
+                             std::vector<ir::Node *> ops) const {
+    return false;
+  }
+
+  bool DealWithSpecialOp(ir::Graph *result, ir::Node *node) const override {
+    if (node->Op()->Type() == "recv") {
+      VLOG(1) << "set recv op do_not_run to true";
```
Comment (seiriosPlus) [inline paddle/fluid/framework/details/multi_devices_graph_pass.h:130]: log level may be higher.

File: paddle/fluid/framework/details/multi_devices_graph_pass.h [hunk 2]
```diff
@@ -115,6 +115,32 @@ class AllReduceSSAGraphBuilder : public MultiDevSSAGraphBuilderBase {
...
+      node->Op()->SetAttr("do_not_run", true);
+      node->Op()->Flush();
+    } else if (node->Name() == "lookup_table" || node->Name() == "nce" ||
```
Comment (seiriosPlus) [inline paddle/fluid/framework/details/multi_devices_graph_pass.h:133]: hard code are not recommend.

File: paddle/fluid/framework/scope.h [hunk 3]
```diff
@@ -52,6 +52,8 @@ class Scope {
   /// Mark it to const because that new kid scope cannot change parent scope.
   Scope& NewScope() const;
 
+  Scope* NewTmpScope() const;
```
Comment (seiriosPlus) [inline paddle/fluid/framework/scope.h]: what's the meaning of ```NewTmpScope ```?

File: paddle/fluid/operators/distributed/parameter_send.cc [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+      VLOG(3) << "sending " << send_var_name << " to " << endpoint;
+      rets.push_back(rpc_client->AsyncSendVar(endpoint, cpu_ctx, *local_scope,
+                                              send_var_name));
+    } else {
+      VLOG(3) << "don't send non-initialized variable: "
+              << rpc_ctx.splited_var_names[i];
+    }
+  }
+
+  // note!! only support sync send now
```
Comment (seiriosPlus) [inline paddle/fluid/operators/distributed/parameter_send.cc]: add Todo.

Comment (jacquesqiao) [reply paddle/fluid/operators/distributed/parameter_send.cc]: fixed

File: paddle/fluid/framework/scope.h (hunk 3 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/scope.h]: Already add some comments.

File: paddle/fluid/framework/details/multi_devices_graph_pass.h (hunk 2 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/details/multi_devices_graph_pass.h:133]: This is not hard code, this method `DealWithSpecialOp` is used to change some special node in the graph like send/recv.

File: paddle/fluid/framework/details/multi_devices_graph_pass.h (hunk 1 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/details/multi_devices_graph_pass.h:130]: This log will only be called in the init phase of async ssa graph ssa graph executor. Used to make sure it work properly.

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc [hunk 5]
```diff
@@ -0,0 +1,203 @@
+//   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/async_ssa_graph_executor.cc:1]: 2018 -> 2019

File: paddle/fluid/framework/parallel_executor.cc [hunk 6]
```diff
@@ -356,21 +403,27 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
...
   }
 
-  member_->executor_.reset(new details::ScopeBufferedSSAGraphExecutor(
-      exec_strategy, member_->local_scopes_, std::move(var_infos),
-      member_->places_, std::move(member_->executor_)));
+  VLOG(3) << "use ScopeBufferedSSAGraphExecutor";
+  if (!build_strategy.async_mode_) {
+    member_->executor_.reset(new details::ScopeBufferedSSAGraphExecutor(
+        exec_strategy, member_->local_scopes_, std::move(var_infos),
+        member_->places_, std::move(member_->executor_)));
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:420]: Why did you not use `ScopeBufferedSSAGraphExecutor` in `async_mode`?

File: paddle/fluid/framework/details/execution_strategy.h [hunk 7]
```diff
@@ -31,6 +31,8 @@ struct ExecutionStrategy {
   size_t num_iteration_per_drop_scope_{1};
   ExecutorType type_{kDefault};
   bool dry_run_{false};
+  size_t num_iteration_per_run_{1};  // only use with async_ssa_graph_executor
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/execution_strategy.h:34]: What's this mean?

File: paddle/fluid/framework/parallel_executor.cc [hunk 8]
```diff
@@ -218,6 +219,18 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
     }
   }
 
+  std::vector<ir::Graph *> graphs;
+  if (build_strategy.async_mode_) {
+    PADDLE_ENFORCE(!member_->use_cuda_,
+                   "gpu mode does not support async_mode_ now!");
+    graphs.push_back(graph);
+    for (int i = 1; i < places.size(); ++i) {
+      auto *tmp_graph = new ir::Graph(graph->OriginProgram());
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:228]: The `OriginProgram` may be a bit different from the current graph.

File: paddle/fluid/framework/parallel_executor.cc [hunk 9]
```diff
@@ -294,19 +307,46 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
...
+    graph = build_strategy.Apply(graph, {member_->places_[0]}, loss_var_name,
+                                 {member_->local_scopes_[0]}, 1,
+                                 member_->use_cuda_, member_->nccl_ctxs_.get());
+    for (int i = 1; i < member_->places_.size(); ++i) {
+      graphs[i] =
+          build_strategy.Apply(graphs[i], {member_->places_[i]}, loss_var_name,
+                               {member_->local_scopes_[i]}, 1,
                                member_->use_cuda_, member_->nccl_ctxs_.get());
+      async_graphs[i] = graphs[i];
+    }
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:327]: Why did you process `i = 0` and `i > 0` respectively here?

File: paddle/fluid/framework/parallel_executor.cc [hunk 10]
```diff
@@ -294,19 +307,46 @@ ParallelExecutor::ParallelExecutor(const std::vector<platform::Place> &places,
...
-// ncclOp
+  // Step 2. Convert main_program to SSA form and dependency graph. Also, insert
+  // ncclOp
+  std::vector<ir::Graph *> async_graphs(places.size());
 #if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)
-  graph = build_strategy.Apply(graph, member_->places_, loss_var_name,
-                               member_->local_scopes_, member_->nranks_,
+  if (build_strategy.async_mode_) {
+    VLOG(3) << "use local async mode";
+    graph = build_strategy.Apply(graph, {member_->places_[0]}, loss_var_name,
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:318]: As line 224 says that `gpu mode does not support async_mode_ now!`, so why did you add code in `#if defined(PADDLE_WITH_CUDA) && !defined(_WIN32)`?

File: paddle/fluid/framework/parallel_executor.cc [hunk 11]
```diff
@@ -425,14 +478,22 @@ void ParallelExecutor::BCastParamsToDevices(
...
+
+        // FIXME(zcd): LR_DECAY_COUNTER should not be shared. This is a hot fix.
+        if (member_->build_strategy_.async_mode_) {
+          share_memory();
+        } else if (member_->use_all_reduce_ || member_->use_cuda_ ||
+                   var == "@LR_DECAY_COUNTER@") {
+          copy_memory();
         } else {
-          t->ShareDataWith(main_tensor);
+          share_memory();
```
Comment (chengduoZH) [inline paddle/fluid/framework/parallel_executor.cc:496]: Why did you make the above modification?

File: paddle/fluid/framework/parallel_executor.cc (hunk 6 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/parallel_executor.cc:420]: ScopeBufferedSSAGraphExecutor will slow done async_ssa_graph_executor and cause many problem.

File: paddle/fluid/framework/details/execution_strategy.h (hunk 7 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/details/execution_strategy.h:34]: https://github.com/PaddlePaddle/Paddle/pull/16172/files#diff-bcb7058cf667aba60603c4448e6180c8R131
used here, will run multi steps when call exe.run to improve performance.

File: paddle/fluid/framework/parallel_executor.cc (hunk 8 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/parallel_executor.cc:228]: What different?

File: paddle/fluid/framework/parallel_executor.cc (hunk 9 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/parallel_executor.cc:327]: This code can be optimized later.

**Review by seiriosPlus — DISMISSED** [review]
LGTM！

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+#ifdef PADDLE_WITH_DISTRIBUTE
+  using RpcCtxMap = operators::distributed::RpcCtxMap;
+  VLOG(3) << "ProcessGraph";
+  RpcCtxMap send_varname_to_ctx;
+  RpcCtxMap recv_varname_to_ctx;
+  for (auto i = 0; i < graphs.size(); ++i) {
+    std::vector<ir::Node *> nodes_to_delete;
+    for (auto &node : graphs[i]->Nodes()) {
+      VLOG(3) << "node name " << node->Name();
+      if (node && node->IsOp()) {
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/async_ssa_graph_executor.cc:58]: Why the node maybe `nullptr` here?

File: paddle/fluid/framework/parallel_executor.cc (hunk 11 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/parallel_executor.cc:496]: For better understanding and code share.

File: paddle/fluid/framework/details/async_ssa_graph_executor.cc (hunk 12 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/details/async_ssa_graph_executor.cc:58]: I have met this problem before, so I add more check to ensure it works right.

File: paddle/fluid/framework/parallel_executor.cc (hunk 10 above)
Comment (jacquesqiao) [reply paddle/fluid/framework/parallel_executor.cc:318]: Will support in the future. This code is used to make sure the compiling works right.

File: paddle/fluid/framework/scope.cc [hunk 13]
```diff
@@ -59,6 +59,8 @@ Scope& Scope::NewScope() const {
   return *child;
 }
 
+Scope* Scope::NewTmpScope() const { return new Scope(this); }
```
Comment (chengduoZH) [inline paddle/fluid/framework/scope.cc]: `NewTmpScope ` should return a `unique_ptr` at least.

Comment (jacquesqiao) [reply paddle/fluid/framework/scope.cc]: Done

**Review by chengduoZH — APPROVED** [review]
Approve for the changing of scope.cc.

File: paddle/fluid/framework/details/async_ssa_graph_executor.h [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+struct VarInfo {
+  std::string name_;
+  proto::VarType::Type type_;
+  bool persistable_;
+};
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/async_ssa_graph_executor.h:33]: Duplicate with [here](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/framework/details/scope_buffered_ssa_graph_executor.h#L35).

**Comment by chengduoZH** [comment]
It is suggested that the code should be polished.
